### PR TITLE
Upgrade tsickle to TS2.5

### PR DESCRIPTION
**Pull request by rkirov**
The major change is that NodeArray is now readonly, but we still
want to modify it in a number of places.

### CODE REVIEW

**Comment by rkirov** [comment]
Not planning to merge this yet, but up for review to see if someone has better ways of working through the API changes. I mostly just added `any`s due to lack of understanding of the code.

**Comment by mprobst** [comment]
Regarding the NodeArray, they must have a pattern they use for their own decorators that still allows manipulating node arrays, right? Or maybe we should re-arrange the code so that we only piece the whole array together at the very end?

/CC @tbosch

**Comment by tbosch** [comment]
Agree, we should just create a new `NodeArray` at the end.

**Comment by tbosch** [comment]
FYI, when using tsickle with TS 2.5 (without the changes in this PR), users get broken compilations. See e.g. https://github.com/angular/angular/issues/19587.

**Comment by tbosch** [comment]
Note that for Angular, as we already committed to supporting TS 2.4, we would need a version of tsickle that works in TS 2.4 and TS 2.5.

**Comment by rkirov** [comment]
Updated the PR with '.slice()' to get a mutable copy of NodeArrays. I didn't see any other convenience APIs for creating new readonly arrays with added/deleted members.

Talked with Chuck about supporting 2.5 and 2.4 through shim. There will be a separate PR for that. When I land this PR, I will create a branch for 2.4 at the current state that we can use to backport patches if needed.

File: src/transformer_util.ts [hunk 1]
```diff
@@ -392,18 +391,19 @@ function visitNodeStatementsWithSynthesizedComments<T extends ts.Node>(
   const leading = synthesizeDetachedLeadingComments(sourceFile, node, statements);
   const trailing = synthesizeDetachedTrailingComments(sourceFile, node, statements);
   if (leading.commentStmt || trailing.commentStmt) {
-    statements = ts.setTextRange(ts.createNodeArray(statements), {pos: -1, end: -1});
```
Comment (chuckjaz) [inline src/transformer_util.ts]: How about something like:
```ts
    const newStatements = [...arrayOf(leading.commentStmt), ...statements, ...arrayOf(trailing.commendStmt)];
```
where
```ts
function arrayOf<T>(value: T | undefined): T[] {
  return value ? [value] : [];
}
```

Comment (rkirov) [reply src/transformer_util.ts]: done

**Comment by rkirov** [comment]
I got the tests to pass with only one questionable assignment.

File: src/class_decorator_downlevel_transformer.ts [hunk 2]
```diff
@@ -96,13 +96,25 @@ function isNameEqual(classMember: ts.ClassElement, name: string): boolean {
...
+    decoratorMetadata: ts.PropertyDeclaration): ts.NodeArray<ts.ClassElement> {
   let insertionPoint = classMembers.findIndex(
       m => isNameEqual(m, 'ctorParameters') || isNameEqual(m, 'propDecorators'));
   if (insertionPoint === -1) {
     insertionPoint = classMembers.length - 1;
   }
-  classMembers.splice(insertionPoint, 0, decoratorMetadata);
+  const members: ts.ClassElement[] = [];
+
+  for (let i = 0; i < classMembers.length;) {
```
Comment (mprobst) [inline src/class_decorator_downlevel_transformer.ts]: what about `members = [classMembers.slice(0, insertionPoint)..., decoratorMetadata, classMembers.slice(insertionPoint)]`?

File: src/transformer_util.ts [hunk 3]
```diff
@@ -288,7 +288,8 @@ export function visitNodeWithSynthesizedComments<T extends ts.Node>(
  * @param node
  */
 function resetNodeTextRangeToPreventDuplicateComments<T extends ts.Node>(node: T): T {
-  ts.setEmitFlags(node, (ts.getEmitFlags(node) || 0) | ts.EmitFlags.NoComments);
+  // tslint:disable-next-line:no-any TS 2.5 typescript.d.ts does not expose getEmitNodes.
```
Comment (mprobst) [inline src/transformer_util.ts]: typo getEmitFlags

File: src/transformer_util.ts [hunk 4]
```diff
@@ -301,11 +302,9 @@ function resetNodeTextRangeToPreventDuplicateComments<T extends ts.Node>(node: T
   if (node.kind === ts.SyntaxKind.PropertyDeclaration) {
     allowTextRange = false;
     const pd = node as ts.Node as ts.PropertyDeclaration;
-    // TODO(tbosch): Using pd.initializer! as the typescript typings before 2.4.0
-    // are incorrect. Remove this once we upgrade to TypeScript 2.4.0.
     node = ts.updateProperty(
-               pd, pd.decorators, pd.modifiers, resetTextRange(pd.name) as ts.PropertyName, pd.type,
-               pd.initializer!) as ts.Node as T;
+               pd, pd.decorators, pd.modifiers, resetTextRange(pd.name) as ts.PropertyName,
+               pd.questionToken, pd.type, pd.initializer) as ts.Node as T;
```
Comment (mprobst) [inline src/transformer_util.ts]: does the 2.4 updateProperty accept the questionToken here?

File: src/transformer_util.ts [hunk 5]
```diff
@@ -392,13 +395,17 @@ function visitNodeStatementsWithSynthesizedComments<T extends ts.Node>(
...
-    }
-    if (trailing.commentStmt) {
-      statements.push(trailing.commentStmt);
-    }
+    const newStatements: ts.Statement[] =
+        [...arrayOf(leading.commentStmt), ...statements, ...arrayOf(trailing.commentStmt)];
+    statements = ts.setTextRange(ts.createNodeArray(newStatements), {pos: -1, end: -1});
+
+    // The visitor creates a new node with the new statements. However, doing so
+    // reveals a TypeScript bug.
```
Comment (mprobst) [inline src/transformer_util.ts]: reference the bug/issue, or at least explain what the bug behaviour is here?

File: src/transformer_util.ts [hunk 6]
```diff
@@ -392,13 +395,17 @@ function visitNodeStatementsWithSynthesizedComments<T extends ts.Node>(
...
+    // The line below is a workaround that ensures that updateSourceFileNode and
+    // updateBlock never create new Nodes.
+    // TODO: narrow this down and file a bug with TS team.
```
Comment (mprobst) [inline src/transformer_util.ts]: TODO format plz

File: src/transformer_util.ts [hunk 7]
```diff
@@ -604,6 +611,7 @@ export function visitEachChild(
  * This is a version of `ts.updateSourceFileNode` that works
  * well with property decorators.
  * See https://github.com/Microsoft/TypeScript/issues/17384
+ * TODO: This has been fixed in TS 2.5. Investigate removal.
```
Comment (mprobst) [inline src/transformer_util.ts]: TODO format

File: src/transformer_util.ts (hunk 4 above)
Comment (rkirov) [reply src/transformer_util.ts]: This change is not meant to be backwards compatible with 2.4. I am confused why are you asking about 2.4?

File: src/class_decorator_downlevel_transformer.ts (hunk 2 above)
Comment (rkirov) [reply src/class_decorator_downlevel_transformer.ts]: done

File: src/transformer_util.ts (hunk 3 above)
Comment (rkirov) [reply src/transformer_util.ts]: done

File: src/transformer_util.ts (hunk 4 above)
Comment (rkirov) [reply src/transformer_util.ts]: As discussed, this change does not have backwards compatibility with TS2.4. I will update the 2.4 branch to the current head before submitting this. Chuck will work on making tsickle 2.4 + 2.5 compatible separately.

File: src/transformer_util.ts (hunk 5 above)
Comment (rkirov) [reply src/transformer_util.ts]: I paired with @tbosch on this comment, already forgot what exactly was breaking (other than tests failing). Tobias, can you help me expanding on the comment?

File: src/transformer_util.ts (hunk 7 above)
Comment (rkirov) [reply src/transformer_util.ts]: done

File: src/transformer_util.ts (hunk 6 above)
Comment (rkirov) [reply src/transformer_util.ts]: done

File: src/transformer_util.ts (hunk 5 above)
Comment (rkirov) [reply src/transformer_util.ts]: Paired with Tobias more today. Still not ideal and somewhat mysterious but we narrowed it down to a small repro case.

**Pull request merged**